### PR TITLE
Open multiple things with a single command

### PR DESCRIPTION
**Pull request by Arxcis**
Implements this behaviour:
multiopen

## TODO
- [x] Add tests
- [x] Add to help menu
- [x] Refactor everything

### CODE REVIEW

File: goto/commands/open.py [hunk 1]
```diff
@@ -28,4 +28,4 @@ def open(magic, args):
     except subprocess.CalledProcessError as e:
         return None, GotoError("open_failed", message=e.message)
 
-    return None, None
+    return "Opened folder: %s" % url, None
```
Comment (technocake) [inline goto/commands/open.py]: open can also open files, not only folders. 

Try it with a textfile or a pdf for instance ^^

**Review by technocake — DISMISSED** [review]
To merge I think we need two things:

1. **Change output of default open:**  
it should not be tied to folders, it may open files and lots of other things (some work already now, try `spotify://something` for instance) .
2. **Add the sexy gif to the readme** under a new section called #Features

**Comment by technocake** [comment]
Oh, the autocompletions should probably be updated to support this behavior also.  It prompts for a slightly more inteligent search layer.  Exciting times :)

**Comment by technocake** [comment]
And one more issue to look put for: The editor launch commands, they will not work with this implementation.

File: goto/commands/open.py (hunk 1 above)
Comment (Arxcis) [reply goto/commands/open.py]: Yes, or .gif, or .png

How about?
```suggestion
    return "Opening: %s" % url, None
```

**Comment by Arxcis** [comment]
Checked
- [x] **Change output of default open:**

- [x] **Add the sexy gif to the readme** under a new section called #Features

**Comment by technocake** [comment]
Well played :) 

Opened a relevant PR to this one: #86  to add verbose mode to the command

Have you been experiencing different behaviour on `webbrowser.open` and `webbrowser.open_new_tab`?  On my tests both are opening a new tab in the browser..

File: goto/commands/open.py (hunk 1 above)
Comment (technocake) [reply goto/commands/open.py]: Perfecto!

**Comment by technocake** [comment]
Is this ready for review now you think @Arxcis ?

**Comment by Arxcis** [comment]
> Is this ready for review now you think @Arxcis ?

Still have to write tests and update the help menu

**Comment by Arxcis** [comment]
> Have you been experiencing different behaviour on `webbrowser.open` and `webbrowser.open_new_tab`? On my tests both are opening a new tab in the browser..

No. The behaviour of these two functions was the same during my tests.

File: README.md [hunk 2]
```diff
@@ -50,7 +50,7 @@ your rcfile should do the trick in most cases.
 
 ```
     The basics
-        goto <magicword>                        Go to shortcut
+        goto [<magicword>...]                   Go to one or more shortcuts
```
Comment (technocake) [inline README.md]: As this is the most central feature of Goto, I think we ought to devote two lines in the help text for it.
One for a single magicword, and one for multiple.

That would  make both examples more clear I believe

File: goto/commands/cd.py [hunk 3]
```diff
@@ -3,7 +3,7 @@
 from ..gotomagic.text import GotoError, GotoWarning
 
 
-def cd(magic, args):
+def cd(magic, command, args):
```
Comment (technocake) [inline goto/commands/cd.py]: 👍  for consistency

File: goto/commands/default.py [hunk 4]
```diff
@@ -4,18 +4,32 @@
 from .open import open
 
 
-def default(magic, magicword):
+def default(magic, command, args):
```
Comment (technocake) [inline goto/commands/default.py]: command should be `None` in this case. 
It is a bit confusing to concatenate args and command.

File: goto/commands/open.py [hunk 5]
```diff
@@ -28,4 +28,4 @@ def open(magic, args):
     except subprocess.CalledProcessError as e:
         return None, GotoError("open_failed", message=e.message)
 
-    return None, None
+    return "%s" % url, None
```
Comment (technocake) [inline goto/commands/open.py]: 1. the % syntax here is unecessary, could just use `url`
2. I would want this output to be shown only when `-v` or `--verbose` is added to the command
In order to keep consistent behaviour.  
`verbose = '-v' in args or '--verbose' in args`
`output = output if verbose else None`

File: goto/commands/usage.py [hunk 6]
```diff
@@ -10,7 +10,7 @@ def usage():
         project help                  Consult my brother in command
 
     The basics
-        goto <magicword>                        Go to shortcut
+        goto [<magicword>...]                   Go to one or more shortcuts
```
Comment (technocake) [inline goto/commands/usage.py]: See comment on README.md

File: goto/tests/test_end_to_end.sh [hunk 7]
```diff
@@ -217,26 +217,57 @@ function test_06_goto_add {
...
-    _cmd_should_succeed "goto add $magicword $uri"
-    _cmd_should_fail "goto $nonexisting_magicword"
-    _failing_cmd_should_give_human_message "goto $nonexisting_magicword"
-    _failing_cmd_should_not_print_ah_hoy_twice "goto $nonexisting_magicword"
+    for i in "${!magicwords[@]}"; do
+        _projectfile_should_contain "${magicwords[$i]}";
+    done
+
+    for i in "${!magicwords[@]}"; do
+        _cmd_should_succeed "goto ${magicwords[$i]}";
```
Comment (technocake) [inline goto/tests/test_end_to_end.sh]: Wow - it is actually a little worrying that these commands succeed now, as `goto+subl` is not a registered scheme on any computer at the time of this writing.

This is an old bug or misbehaving feature of goto if you want to call it that

File: goto/tests/test_end_to_end.sh [hunk 8]
```diff
@@ -217,26 +217,57 @@ function test_06_goto_add {
...
+    declare -a uris
+    magicwords=(
+        "red"
+        "blue"
+        "green"
+        "yellow"
+        "orange"
+    )
+    uris=(
+        "\"http://red.com\""
```
Comment (technocake) [inline goto/tests/test_end_to_end.sh]: I do indeed see the need to test multiple schemes. 
However I think we should think a little more about using these schemes in the automated tests for now ..
They should work in Circle Ci, which runs them in a linux docker container (and a linux VM) without any desktop.

They will causesideeffects from running the tests - opening windows of different kinds.
I'm not sure they would all work in different environments. 

And this also binds goto developers to have spotify installed -- I kinda don't want to tie goto to any particular other software

File: goto/tests/test_end_to_end.sh [hunk 9]
```diff
@@ -217,26 +217,57 @@ function test_06_goto_add {
...
+        "\"mailto:jon@gmail.com\""
+        "\"spotify:green.com\""
+        "\"sftp:hallo@yellow.com\""
+        "\"goto+subl://orange.com\""
+    )
+
+    for i in "${!magicwords[@]}"; do
```
Comment (technocake) [inline goto/tests/test_end_to_end.sh]: The use of these for loops is clean!

File: goto/the_real_goto.py [hunk 10]
```diff
@@ -51,42 +51,42 @@ def run_command(magic, command, args):
...
 
     if command == 'cd':
-        return commands.cd(magic, args)
+        return commands.cd(magic, command, args)
 
     if command in ['mv', 'rename']:
         return commands.rename(magic, command, args)
 
-    return commands.default(magic, command)
+    return commands.default(magic, command, args)
```
Comment (technocake) [inline goto/the_real_goto.py]: Ref the comment in `default.py` I believe this should be treated a special case, where the outcome is:

set command to `None` 
and send all sys.argv in the args parameter (except the jfile and scriptname)

**Review by technocake — CHANGES_REQUESTED** [review]

File: goto/tests/test_end_to_end.sh (hunk 7 above)
Comment (technocake) [reply goto/tests/test_end_to_end.sh]: It might be considered outside the scope of this PR to fix the exit status of goto <magicwords> that should fail

File: README.md (hunk 2 above)
Comment (Arxcis) [reply README.md]: Good idea

File: goto/commands/cd.py (hunk 3 above)
Comment (Arxcis) [reply goto/commands/cd.py]: YEs. Can't believe I missed that one

File: goto/commands/default.py (hunk 4 above)
Comment (Arxcis) [reply goto/commands/default.py]: Ahh, of course. On it

File: goto/tests/test_end_to_end.sh (hunk 8 above)
Comment (Arxcis) [reply goto/tests/test_end_to_end.sh]: Alright. I didn't know the tests actually opens windows.

Comment (technocake) [reply goto/tests/test_end_to_end.sh]: Don't misunderstand, these are powerful tests. 
It might be a possibility to run this test conditionally - if explicitly setting an environment wariable or a command line argument. 

The side-effect of opening multiple windows should be optional  on devs machines -- and I hope we can find a viable solution for making  this run and verify the results in  Circle CI. 

You kinda hit a nerve here -- or an effect of goto that is really important to test - but hard to test within the current framework without causing side-effects and making sure the tests will run on different environments.

File: goto/the_real_goto.py (hunk 10 above)
Comment (Arxcis) [reply goto/the_real_goto.py]: How about
```py
    args = [command] + args
    return commands.default(magic, args)
```
vs
```py
    command = None
    return commands.default(magic, command, args)
```
I like the first one better

File: goto/commands/cd.py (hunk 3 above)
Comment (technocake) [reply goto/commands/cd.py]: You didn't miss it, I just wanted to point out that I love that all the commands have the same parameter signature now :)

File: goto/tests/test_end_to_end.sh (hunk 8 above)
Comment (technocake) [reply goto/tests/test_end_to_end.sh]: Actually, they should open windows, if they worked ^^

File: goto/tests/test_end_to_end.sh (hunk 9 above)
Comment (Arxcis) [reply goto/tests/test_end_to_end.sh]: I agree, if you don't consider the monster in the middle which is `"${!magicwords[@]}"`

Comment (Arxcis) [reply goto/tests/test_end_to_end.sh]: "bash-clean"

File: goto/tests/test_end_to_end.sh (hunk 8 above)
Comment (Arxcis) [reply goto/tests/test_end_to_end.sh]: `make test` has never opened a window for me

File: goto/the_real_goto.py (hunk 10 above)
Comment (technocake) [reply goto/the_real_goto.py]: First one seems to be cleaner perhaps if adding an explaining comment

File: goto/the_real_goto.py [hunk 11]
```diff
@@ -10,18 +10,17 @@
 from .gotomagic import text
 from .gotomagic.magic import GotoMagic
 from .gotomagic.utils import healthcheck
-
 from . import commands
 
-# make sure we print in utf-8
```
Comment (technocake) [inline goto/the_real_goto.py:19]: Why is this removed?

File: goto/tests/test_end_to_end.sh (hunk 8 above)
Comment (Arxcis) [reply goto/tests/test_end_to_end.sh]: > You kinda hit a nerve here -- or an effect of goto that is really important to test - but hard to test within the current framework without causing side-effects and making sure the tests will run on different environments.

I thought for a second about "maybe I should not do this?", but then some other person in my head said "You should do whatever you want. It's a test! It is meant to break things. Make a mess!"

File: goto/the_real_goto.py [hunk 12]
```diff
@@ -10,18 +10,17 @@
...
-try:
-    if sys.stdout.encoding != 'utf-8':
-        sys.stdout = codecs.getwriter('utf-8')(sys.stdout.buffer, 'strict')
-except:
-    pass  # TODO: implement utf-8 encoding of py2.7
-
 
 def main():
+    # make sure we print in utf-8
```
Comment (technocake) [inline goto/the_real_goto.py]: oh I see why - it was moved. not removed.
Makes sense.

Would be cleaner to have it in it's own function somwhere I guess. -- but that is outside the scope of this PR - I'll fix it in the unicode branch

File: goto/the_real_goto.py [hunk 13]
```diff
@@ -45,6 +44,8 @@ def main():
         print(output)
         exit(0)
 
+    exit(0)
```
Comment (technocake) [inline goto/the_real_goto.py:72]: `explicit is better then implicit`

Good catch

**Review by technocake — APPROVED** [review]

File: goto/tests/test_end_to_end.sh (hunk 9 above)
Comment (technocake) [reply goto/tests/test_end_to_end.sh]: Hehe - well goto is a project that could level up bash skills.

**Review by technocake — APPROVED** [review]

File: goto/commands/cd.py (hunk 3 above)
Comment (Arxcis) [reply goto/commands/cd.py]: Oooo.. Yes I see now that it is consistent. Which was the intent.

Maybe the `cd` command needs to know the command which initiated it, maybe not. The argparser should not care, except for the default edge-case.

File: goto/tests/test_end_to_end.sh (hunk 8 above)
Comment (technocake) [reply goto/tests/test_end_to_end.sh]: hehe - I think that if no windows are opened - it is actually a symptom of a misbehaving goto. 
Then goto should `exit 1` for sure.

File: goto/tests/test_end_to_end.sh (hunk 9 above)
Comment (technocake) [reply goto/tests/test_end_to_end.sh]: I must admit I had to google:
https://www.linuxjournal.com/content/bash-arrays

File: goto/commands/cd.py (hunk 3 above)
Comment (technocake) [reply goto/commands/cd.py]: Actually ... the cd command is never used... 

The cd is done in the `shell-script` layer, before it even hits the python layer. 

That osascript leftovers there, is an early experiment I did on cd-ing, that would only work on osx and it would also open a new terminal window.  

It is safe to delete this command.py

File: goto/tests/test_end_to_end.sh (hunk 8 above)
Comment (technocake) [reply goto/tests/test_end_to_end.sh]: I think we consider fixing this as outside the scope of this PR - as it is really two issues:

 1. an issue with goto on failing open/start calls. 
2. an issue with the test system that should actually cause side effects if calling these commands.

File: goto/tests/test_end_to_end.sh (hunk 9 above)
Comment (Arxcis) [reply goto/tests/test_end_to_end.sh]: Thats a really good resource. Best I have seen. Always have to google how the bash arrays work. They are a bit .... different

File: goto/the_real_goto.py [hunk 14]
```diff
@@ -34,9 +34,13 @@ def main():
     jfile = sys.argv[1]
     magic = GotoMagic(jfile)
     command = sys.argv[2]
-    args = sys.argv[3:]
+    argv = sys.argv[3:]
 
-    output, err = run_command(magic, command, args)
+    args = list(filter(lambda word: not word.startswith('-'), argv))
```
Comment (Arxcis) [inline goto/the_real_goto.py:55]: I see now that this will filter the "commands" `-o` and `--open`

Comment (technocake) [reply goto/the_real_goto.py:55]: yes - I suppose those should be treated as **options** and sent to the commands as the options list

Comment (Arxcis) [reply goto/the_real_goto.py:55]: Ideally we should require the following order on all commands:
```
goto <command> [<options>...] [<args>...]
```

Quite close to the convention of POSIX utilities described here https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap12.html

Comment (Arxcis) [reply goto/the_real_goto.py:55]: Or whatever order people prefer
```
goto <option> <command> <option> <option> <arg> <arg> <arg> <option>

Comment (technocake) [reply goto/the_real_goto.py:55]: `goto add existing new-url` -->  Ah hoy!  ...  `goto add existing new-url -f`
`goto rename existing1 existing2` --> Ah hoy!  ...  ↑ `goto add existing1 existing2 -f` -->  Renamed.

Comment (technocake) [reply goto/the_real_goto.py:55]: Yes - by the time we want to support values in options of the type:  `--option value` -- we are far better off at putting our effort into an existing cli arg parsing framework like click, fire or even the builtin argparse.

Comment (Arxcis) [reply goto/the_real_goto.py:55]: Maybe if we keep it simple enough, we won't need complicated option handling.

**Comment by Arxcis** [comment]
>goto add existing new-url --> Ah hoy! ... goto add existing new-url -f
goto rename existing1 existing2 --> Ah hoy! ... ↑ goto add existing1 existing2 -f --> Renamed.

This does not work currently right?

**Comment by Arxcis** [comment]
Waiting for #107 to me merged before continuing work on this PR

**Comment by technocake** [comment]
Want some assistance on merging them?

**Comment by Arxcis** [comment]
> @Arxcis: want some assistance on merging 107 and this branch?

I am just thinking merging like this
```
107 -> develop
develop -> 76
76 -> develop
```

**Comment by technocake** [comment]
> > @Arxcis: want some assistance on merging 107 and this branch?
> 
> I am just thinking merging like this
> 
> ```
> 107 -> develop
> develop -> 76
> 76 -> develop
> ```

Yeah, it was more the effort involved I was thinking about.  Oki then, enjoy ^^

**Comment by Arxcis** [comment]
> Yeah, it was more the effort involved I was thinking about. Oki then, enjoy ^^

When we meet up, we could do the merging collaboratively 👍

**Comment by technocake** [comment]
Overview of all args and options leading to a command is here:
https://github.com/technocake/goto/wiki/Command_Parsing

File: README.md [hunk 15]
```diff
@@ -37,9 +37,15 @@ Do the same as above, but **open git bash as Administrator**
 #### Linux - User Site install
 On linux, the recommended way to install goto is by `pip install --user magicgoto`
 
+<<<<<<< HEAD
+This requires that you have your User site bin  in your path.
```
Comment (technocake) [inline README.md]: oops

Comment (technocake) [reply README.md]: Fixed in  commit 966369283477baa431af4302e9a2487fd1e7e4dc
